lib/semian/net_http.rb

```diff
@@ -87,6 +87,16 @@ def disabled?
       raw_semian_options.nil?
     end
 
+    def acquire_semian_resource(*)
+      result = super
+
+      if result.is_a?(::Net::HTTPServerError) && raw_semian_options.fetch(:open_circuit_server_errors)
```

Comment:
That was just a stopping point after a while of looking at this, that's not what I think this should look like. 

I started out with something simple like this: 

```
    def acquire_semian_resource(*)
      result = super

      if result.is_a?(::Net::HTTPServerError) && raw_semian_options.fetch(:open_circuit_server_errors)
        semian_resource.mark_failed(result)
      end

      result
    end
```

But this started failing the test I wrote (the one that tests that 500s open the circuit). 

It appeared that each [`http.get("/500")`](https://github.com/Shopify/semian/pull/149/files#diff-416f865ec7d8b97ae535d742b47da46eR333) would reach `semian_resource.mark_failed(result)` in the snippet above _twice_. Causing the circuit to be opened prematurely _before_ the `error_threshold` is reached. 

This happens because `raw_request` calls our `request` (and `connect` as well!) method again if there isn't a connection already. See: https://github.com/ruby/ruby/blob/trunk/lib/net/http.rb#L1425

Not sure what the path forwards looks like, we're still thinking about it. One option is: 

```
    def request(req, body = nil, &block)
      return raw_request(req, body, &block) if disabled?
      start if !started?
      acquire_semian_resource(adapter: :http, scope: :query) { raw_request(req, body, &block) }
    end
```
Which would avoid that second `request` invocation from `raw_request`. 

We're still thinking about it, let me know if you have any thoughts in the meantime.